contrib/atf/atf-lua/atf-lua.3
.\"
.\" SPDX-License-Identifier: BSD-2-Clause-FreeBSD
.\"
.\" Copyright (c) 202o Kyle Evans <kevans@FreeBSD.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
.Dd October 21, 2020
.Dt ATF-LUA 3
.Os
.Sh NAME
.Nm atf.TestCase ,
.Nm atf.check_equal ,
.Nm atf.config_get ,
.Nm atf.config_has ,
.Nm atf.expect_death ,
.Nm atf.expect_exit ,
.Nm atf.expect_fail ,
.Nm atf.expect_pass ,
.Nm atf.expect_signal ,
.Nm atf.expect_timeout ,
.Nm atf.fail ,
.Nm atf.get ,
.Nm atf.get_srcdir ,
.Nm atf.pass ,
.Nm atf.require_prog ,
.Nm atf.set ,
.Nm atf.skip
.Nd Lua API to write ATF-based test programs
.Sh SYNOPSIS
.Dv atf.TestCase
.Pp
.Fn atf.check_equal "expected_expression" "actual_expression"
.Fn atf.config_get "varname" "default"
.Fn atf.config_has "varname"
.Fn atf.expect_death "reason"
.Fn atf.expect_exit "reason" "exitcode"
.Fn atf.expect_fail "reason"
.Fn atf.expect_pass
.Fn atf.expect_signal "reason" "signal"
.Fn atf.expect_timeout "reason"
.Fn atf.fail "reason"
.Fn atf.get "varname"
.Fn atf.get_srcdir
.Fn atf.pass
.Fn atf.require_prog "prog_name"
.Fn atf.set "varname" "value"
.Fn atf.skip "reason"
.Sh DESCRIPTION
ATF
provides a simple but powerful interface to easily write test programs in
the Lua scripting language.
These are extremely helpful given that they are trivial to write due to the
language simplicity and the great deal of available external tools, so they
are often ideal to test other applications at the user level.
.Pp
Test programs written using this library must be run using the
.Xr atf-lua 1
interpreter by putting the following on their very first line:
.Bd -literal -offset indent
#! /path/to/bin/atf-lua
.Ed
.Pp
Lua-based test programs are more flexible than their
.Xr atf-sh 1
counterpart.
The basic test structure will resemble this:
.Bd -literal -offset indent
local atf = require('atf')

atf.TestCase "tc1" {
	head = function()
		... first test case's header ...
	end,
	body = function()
		... first test case's body ...
	end,
}

atf.TestCase "tc2" {
	head = function()
		... second test case's header ...
	end,
	body = function()
		... second test case's body ...
	end,
	cleanup = function()
		... second test case's cleanup ...
	end,
}

\&... additional test cases ...
.Ed
.Ss Definition of test cases
Test cases have an identifier and are composed of three different parts:
the header, the body and an optional cleanup routine, all of which are
described in
.Xr atf-test-case 4 .
To define test cases, one must construct it with
.Dv atf.TestCase ,
which takes a first parameter specifying the test case's name and instructs the
library to accept it as a valid test case.
The second parameter is an object defining the test case, which should include
.Fn body ,
and may include
.Fn head
and
.Fn cleanup .
.Pp
It is important to note that an
.Dv atf.TestCase ,
unlike in other
.Xr atf 7
framework libraries,
.Em does
set the test case up for execution when the program is run.
.Ss Program initialization
This library does not have an analog for the
.Nm atf_init_test_cases
required by
.Xr atf-sh 3
to register test cases to run.
.Ss Configuration variables
The test case has read-only access to the current configuration variables
through the
.Nm atf.config_has
and
.Nm atf.config_get
methods.
The former takes a single parameter specifying a variable name and returns
a boolean indicating whether the variable is defined or not.
The latter can take one or two parameters.
If it takes only one, it specifies the variable from which to get the
value, and this variable must be defined.
If it takes two, the second one specifies a default value to be returned
if the variable is not available.
.Ss Access to the source directory
It is possible to get the path to the test case's source directory from
anywhere in the test program by using the
.Nm atf.get_srcdir
function.
.Ss Requiring programs
Aside from the
.Va require.progs
meta-data variable available in the header only, one can also check for
additional programs in the test case's body by using the
.Nm atf.require_prog
function, which takes the base name or full path of a single binary.
Relative paths are forbidden.
If it is not found, the test case will be automatically skipped.
.Ss Test case finalization
The test case finalizes either when the body reaches its end, at which
point the test is assumed to have
.Em passed ,
or at any explicit call to
.Nm atf.pass ,
.Nm atf.fail
or
.Nm atf.skip .
These three functions terminate the execution of the test case immediately.
The cleanup routine will be processed afterwards in a completely automated
way, regardless of the test case's termination reason.
.Pp
.Nm atf.pass
does not take any parameters.
.Nm atf.fail
and
.Nm atf.skip
take a single string parameter that describes why the test case failed or
was skipped, respectively.
It is very important to provide a clear error message in both cases so that
the user can quickly know why the test did not pass.
.Ss Expectations
Everything explained in the previous section changes when the test case
expectations are redefined by the programmer.
.Pp
Each test case has an internal state called
.Sq expect
that describes what the test case expectations are at any point in time.
The value of this property can change during execution by any of:
.Bl -tag -width indent
.It Fn atf.expect_death "reason"
Expects the test case to exit prematurely regardless of the nature of the
exit.
.It Fn atf.expect_exit "reason" "exitcode"
Expects the test case to exit cleanly.
.Pp
.Fa exitcode
is optional.
If
.Fa exitcode
is provided, the runtime engine will validate that the exit code of the test
case matches the one provided in this call.
Otherwise, the exact value will be ignored.
.It Fn atf.expect_fail "reason"
Any failure raised in this mode is recorded, but such failures do not report
the test case as failed; instead, the test case finalizes cleanly and is
reported as
.Sq expected failure ;
this report includes the provided
.Fa reason
as part of it.
If no error is raised while running in this mode, then the test case is
reported as
.Sq failed .
.Pp
This mode is useful to reproduce actual known bugs in tests.
Whenever the developer fixes the bug later on, the test case will start
reporting a failure, signaling the developer that the test case must be
adjusted to the new conditions.
In this situation, it is useful, for example, to set
.Fa reason
as the bug number for tracking purposes.
.It Fn atf.expect_pass
This is the normal mode of execution.
In this mode, any failure is reported as such to the user and the test case
is marked as
.Sq failed .
.It Fn atf.expect_signal "reason" "signo"
Expects the test case to terminate due to the reception of a signal.
.Pp
.Fa signo
is optional.
If
.Fa signo
is provided, the runtime engine will validate that the signal that terminated
the test case matches the one provided in this call.
Otherwise, the exact value will be ignored.
.It Fn atf.expect_timeout "reason"
Expects the test case to execute for longer than its timeout.
.El
.Ss Helper functions for common checks
.Bl -tag -width indent
.It Nm atf.check_equal "expected_expression" "actual_expression"
This function takes two expressions, evaluates them and, if their
results differ, aborts the test case with an appropriate failure message.
The common style is to put the expected value in the first parameter and the
actual value in the second parameter.
.El
.Ss Test inheritance
Test programs written with
.Nm
can take advantage of test object inheritance.
By default, all tests are derived from the
.Dv atf.TestCase
.Dq class
and auto-registered, but test cases can also opt out of auto-registration if
they're primarily intended for being inherited by providing a boolean
.Va atf_auto
in the test case definition.
.Pp
See
.Sx EXAMPLES .
.Sh EXAMPLES
The following shows a complete test program with a single test case that
validates the addition operator:
.Bd -literal -offset indent
local atf = require('atf')

atf.TestCase "addition" {
	head = function()
		atf.set("descr", "Sample tests for the addition operator")
	end,
	body = function()
		atf.check_equal(0, 0 + 0)
		atf.check_equal(1, 0 + 1)
		atf.check_equal(1, 1 + 0)
		atf.check_equal(2, 1 + 1)
		atf.check_equal(300, 100 + 200)
	end,
}
.Ed
.Pp
The following showcases various modes of test inheritance that are allowed:
.Bd -literal -offset indent
local atf = require('atf')

-- The ident string of a non-auto TestCase is generally unused, since these are
-- not auto-registered by default.  They're not technically required to be
-- unique.
local TestSkel = atf.TestCase "skeleton" {
	atf_auto = false,
	-- This head function will be called for any tests that inherit from
	-- TestSkel.  Note that a body is not provided here, so by default a
	-- test derived from this will fail because they're unimplemented.
	head = function()
		atf.set("require.user", "unprivileged")
	end,
}

-- The local we assigned to the result of the above atf.TestCase expression
-- can then be used to derive another test, RequiresUnpriv.
TestSkel "RequiresUnpriv" {
	-- The atf_auto property is not inherited, and it's assumed to be true
	-- if it's not set.  Therefore, any test derived from the above skeleton
	-- will still get autoregistered.
	body = function()
		-- Execute some things that require an unprivileged user.
	end,
}
.Ed
.Sh SEE ALSO
.Xr atf-lua 1 ,
.Xr atf-test-program 1 ,
.Xr atf-test-case 4


contrib/atf/atf-lua/atf-lua.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
 *
 * Copyright (c) 2020 Kyle Evans <kevans@FreeBSD.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include <sys/stat.h>

#include <errno.h>
#include <libgen.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#define linit_c
#define LUA_LIB

#include "lprefix.h"

#include "lua.h"

#include "lualib.h"
#include "lauxlib.h"

#include "latf.h"

#if defined(HAVE_GNU_GETOPT)
#   define GETOPT_POSIX "+"
#else
#   define GETOPT_POSIX ""
#endif

#define	usage_error(...)	usage_error_code(EXIT_FAILURE, __VA_ARGS__)

/*
** these libs are loaded by lua.c and are readily available to any Lua
** program
*/
static const luaL_Reg loadedlibs[] = {
  {"_G", luaopen_base},
  {LUA_LOADLIBNAME, luaopen_package},
  {LUA_COLIBNAME, luaopen_coroutine},
  {LUA_TABLIBNAME, luaopen_table},
  {LUA_IOLIBNAME, luaopen_io},
  {LUA_OSLIBNAME, luaopen_os},
  {LUA_STRLIBNAME, luaopen_string},
  {LUA_MATHLIBNAME, luaopen_math},
  {LUA_UTF8LIBNAME, luaopen_utf8},
  {LUA_DBLIBNAME, luaopen_debug},
#if defined(LUA_COMPAT_BITLIB)
  {LUA_BITLIBNAME, luaopen_bit32},
#endif
  {"atf", luaopen_atf},
  {NULL, NULL}
};

static const char *m_prog_name;

static void
atf_lua_openlibs (lua_State *L) {
	for (const luaL_Reg *lib = loadedlibs; lib->func != NULL; lib++) {
		luaL_requiref(L, lib->name, lib->func, 1);
		lua_pop(L, 1);
	}
}

static void
runtime_error_ap(const char *fmt, va_list ap)
{
	fprintf(stderr, "%s: ERROR: ", m_prog_name);
	vfprintf(stderr, fmt, ap);
	fprintf(stderr, "\n");
}

static void __dead2
runtime_error(int exitcode, const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	runtime_error_ap(fmt, ap);
	va_end(ap);

	exit(exitcode);
}

static void __dead2
usage_error_code(int exitcode, const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	runtime_error_ap(fmt, ap);
	va_end(ap);

	fprintf(stderr, "%s: See atf-lua(1) for usage details.\n",
	    m_prog_name);
	exit(exitcode);
}

static int
atf_lua_panic(lua_State *L)
{
	struct latf_error *error, **errorp;

	/* Not ours, just push it as a string. */
	if (!lua_isuserdata(L, -1)) {
		fprintf(stderr, "Lua error: %s\n", lua_tostring(L, -1));
		exit(1);
	}

	errorp = (struct latf_error **)luaL_checkudata(L, -1,
	    LATF_ERROR_METATABLE);
	error = *errorp;

	if (error->err_msg == NULL)
		exit(error->err_exitcode);

	runtime_error(error->err_exitcode, error->err_msg);
}

static void
atf_lua_parse_tcname(char *tcarg, const char **tcname, const char **tcmethod)
{
	char *sep;

	*tcname = tcarg;
	*tcmethod = NULL;
	sep = strchr(tcarg, ':');
	if (sep == NULL) {
		*tcmethod = "body";
		return;
	}

	*sep++ = '\0';
	*tcmethod = sep;

	if (strcmp(sep, "body") != 0 && strcmp(sep, "cleanup") != 0)
		usage_error("Unknown test case part `%s'", *sep);
}

static int
atf_execute(lua_State *L, char *argv[])
{
	const char *tcname, *tcmethod;
	int ret;

	/* Will exit if m_argv[0] is malformed. */
	atf_lua_parse_tcname(argv[0], &tcname, &tcmethod);

	/*
	 * The latf layer will again bubble any errors straight to the panic
	 * handler that we've setup, with exception to a couple checks that it
	 * makes early on and indicates via the return value.  Currently, the
	 * main error we'll observe upon return is that the test case wasn't
	 * registered.
	 *
	 * This is the easy part, as latf_execute() just needs to invoke the
	 * head() method on whichever test we've been instructed to run, then
	 * run the request method.  Whatever may be driving us, probably kyua,
	 * will drive invocation of the cleanup method if needed -- that still
	 * goes through this path.
	 */
	ret = latf_execute(L, tcname, tcmethod);
	if (ret == 0)
		return (EXIT_SUCCESS);
	else if (ret == ENOENT) {
		usage_error("Unknown test case `%s'", tcname);
	}

	usage_error("Unhandled return/error d", ret);
}

int
main(int argc, char *argv[])
{
	const char *atf_magic, *script, *srcdir, *resultfile;
	char **m_argv;
	lua_State *L;
	int ch, m_argc, ret;
	bool lflag;
	struct stat sb;

	m_prog_name = basename(argv[0]);
	/* libtool workaround: skip the "lt-" prefix if present. */
	if (strncmp(m_prog_name, "lt-", 3) == 0)
		m_prog_name += 3;

	m_argc = --argc;
	m_argv = ++argv;

	if (m_argc < 1)
		usage_error("No test program provided");

	script = m_argv[0];
	if (stat(script, &sb) != 0)
		runtime_error(EXIT_FAILURE, "The test program '%s' does not exist", script);

	L = luaL_newstate();
	if (L == NULL)
		runtime_error(EXIT_FAILURE, "Failed to create state: not enough memory");

	atf_lua_openlibs(L);

	/*
	 * We set the panic handler here, but realistically it's OK to set it
	 * any time after loading libs above and before loading the file below
	 * since there's no need to catch errors in the early stages of
	 * initialization -- none of those will return our custom userdata
	 * error.
	 */
	lua_atpanic(L, atf_lua_panic);
	latf_set_args(L, m_argc, m_argv);

	lflag = 0;
	srcdir = "";
	resultfile = "/dev/stdout";
	while ((ch = getopt(m_argc, m_argv, GETOPT_POSIX ":lr:s:v:")) != -1) {
		switch (ch) {
		case 'l':
			lflag = true;
			break;

		case 'r':
			resultfile = optarg;
			break;

		case 's':
			srcdir = optarg;
			break;

		case 'v':
			if (*optarg == '\0')
				runtime_error(EXIT_FAILURE,
				    "-v requires a non-empty argument");
			if (!latf_add_var(L, optarg))
				runtime_error(EXIT_FAILURE,
				    "-v requires an argument of the form "
				    "var=value");
			break;

		case ':':
			usage_error("Option -%c requires an argument.", optopt);
			break;

		case '?':
		default:
			usage_error("Unknown option -%c.", optopt);
		}
	}
	m_argc -= optind;
	m_argv += optind;

	latf_set_resultfile(L, resultfile);

	/*
	 * srcdir gets plopped into the test config.  One can fetch it with
	 * atf.config_get("srcdir"), but we also provide atf.get_srcdir() for
	 * some consistency with atf-sh(3).
	 */
	latf_set_srcdir(L, srcdir);

	/*
	 * It's important to make sure we're generally ready to start executing
	 * things *before* loading the script here.  Primarily, we want to make
	 * sure the srcdir is intact in case that's somehow relevant to test
	 * case registration.  All registration will happen at this point.
	 */
	ret = luaL_dofile(L, script);
	if (ret != LUA_OK) {
		const char *errstr = lua_tostring(L, -1);

		errstr = errstr == NULL ? "unknown" : errstr;
		runtime_error(EXIT_FAILURE, "Error while executing %s: %s",
		    script, errstr);
	}

	if (lflag) {
		if (m_argc > 0)
		    usage_error("Cannot provide test case names with -l");

		/*
		 * `-l` is an easy one to deal with.  latf_list will iterate
		 * over all the tests that registered when we initially loaded
		 * the script.  Each test will have its head method invoked,
		 * then latf_list will pull and output any vars that were set
		 * in the head() function.
		 */
		latf_list(L);

		/* All errors will hit our panic handler. */
		return (EXIT_SUCCESS);
	}

	if (m_argc == 0)
		usage_error("Must provide a test case name");
	else if (m_argc > 1)
		usage_error("Cannot provide more than one test case name");

	atf_magic = getenv("__RUNNING_INSIDE_ATF_RUN");
	if (atf_magic == NULL || strcmp(atf_magic, "internal-yes-value") != 0) {
		fprintf(stderr, "%s: WARNING: Running test cases outside "
		   "of kyua(1) is unsupported\n", m_prog_name);
		fprintf(stderr, "%s: WARNING: No isolation nor timeout "
		    "control is being applied; you may get unexpected failures; see "
		    "atf-test-case(4)\n", m_prog_name);
	}

	return (atf_execute(L, m_argv));
}
